File: regnet/regnet.py
import os
import sys
import struct
import argparse

import numpy as np
import pycuda.autoinit  # noqa: F401
import pycuda.driver as cuda
import tensorrt as trt

BATCH_SIZE = 1
INPUT_H = 224
INPUT_W = 224
OUTPUT_SIZE = 1000
INPUT_BLOB_NAME = "data"
OUTPUT_BLOB_NAME = "prob"

WEIGHT_PATH = "./models/regnet_x_400mf.wts"
ENGINE_PATH = "./models/regnet_x_400mf.engine"

TRT_LOGGER = trt.Logger(trt.Logger.INFO)


def load_weights(file):
    print(f"Loading weights: {file}")

    assert os.path.exists(file), 'Unable to load weight file.'

    weight_map = {}
    with open(file, "r") as f:
        lines = [line.strip() for line in f]
    count = int(lines[0])
    assert count == len(lines) - 1
    for i in range(1, count + 1):
        splits = lines[i].split(" ")
        name = splits[0]
        cur_count = int(splits[1])
        assert cur_count + 2 == len(splits)
        values = []
        for j in range(2, len(splits)):
            # hex string to bytes to float
            values.append(struct.unpack(">f", bytes.fromhex(splits[j])))
        weight_map[name] = np.array(values, dtype=np.float32)

    return weight_map


def add_batchnorm_2d(network, weight_map, lname, input):
    gamma = weight_map[lname + ".weight"]
    beta = weight_map[lname + ".bias"]
    mean = weight_map[lname + ".running_mean"]
    var = weight_map[lname + ".running_var"]

    eps = 1e-5
    scale = gamma / np.sqrt(var + eps)
    shift = beta - mean * scale
    power = np.ones_like(scale)

    bn = network.add_scale(
        input=input,
        mode=trt.ScaleMode.CHANNEL,
        shift=shift,
        scale=scale,
        power=power,
    )
    return bn


def conv_bn_relu(network, weight_map, lname, input, out_ch, ksize, stride, padding, groups=1):
    # Conv
    conv = network.add_convolution_nd(
        input=input,
        num_output_maps=out_ch,
        kernel_shape=(ksize, ksize),
        kernel=weight_map[lname + ".0.weight"],
        bias=None
    )
    conv.stride_nd = (stride, stride)
    conv.padding_nd = (padding, padding)
    conv.num_groups = groups

    # BN
    bn = add_batchnorm_2d(network, weight_map, lname + ".1", conv.get_output(0))

    # ReLU
    relu = network.add_activation(bn.get_output(0), trt.ActivationType.RELU)

    return relu


def add_resbottleneck_block(network, weight_map, lname, input, in_ch, out_ch, stride, groups):
    # f.a conv shape
    # out_ch * in_ch * kernel_h * kernel_w = size
    # → out_ch = size / (in_ch * k_h * k_w)
    w_a = weight_map[lname + ".f.a.0.weight"]
    oc_a = int(w_a.size / (input.get_output(0).shape[1] * 1 * 1))
    bottleneck = oc_a   # f.a output channel.

    # f.c conv shape
    w_c = weight_map[lname + ".f.c.0.weight"]
    out_ch = int(w_c.size / (bottleneck * 1 * 1))

    # 1. projection branch
    in_tensor = input.get_output(0) if hasattr(input, "get_output") else input
    if stride != 1 or in_ch != out_ch:

        proj_conv = network.add_convolution_nd(
            input=in_tensor,
            num_output_maps=out_ch,
            kernel_shape=(1, 1),
            kernel=weight_map[lname + ".proj.0.weight"],
            bias=None
        )
        proj_conv.stride_nd = (stride, stride)

        proj_bn = add_batchnorm_2d(
            network, weight_map, lname + ".proj.1", proj_conv.get_output(0)
        )
        proj_out = proj_bn.get_output(0)
    else:
        proj_out = in_tensor

    # 2. BottleNeck f.a (1×1)
    a = conv_bn_relu(
        network, weight_map, lname + ".f.a",
        input=input.get_output(0),
        out_ch=bottleneck,
        ksize=1,
        stride=1,
        padding=0
    )

    # 3. f.b (3×3 groups)
    b = conv_bn_relu(
        network, weight_map, lname + ".f.b",
        input=a.get_output(0),
        out_ch=bottleneck,
        ksize=3,
        stride=stride,
        padding=1,
        groups=groups
    )

    # 4. f.c (1×1)
    c_conv = network.add_convolution_nd(
        input=b.get_output(0),
        num_output_maps=out_ch,
        kernel_shape=(1, 1),
        kernel=weight_map[lname + ".f.c.0.weight"],
        bias=None
    )
    c_bn = add_batchnorm_2d(network, weight_map, lname + ".f.c.1", c_conv.get_output(0))

    # 5. ElementWise sum
    ew = network.add_elementwise(
        c_bn.get_output(0),
        proj_out,
        trt.ElementWiseOperation.SUM
    )

    # 6. ReLU
    out = network.add_activation(ew.get_output(0), trt.ActivationType.RELU)
    return out


def create_engine(builder, config, dt):
    weight_map = load_weights(WEIGHT_PATH)

    network = builder.create_network(1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH))

    data = network.add_input(INPUT_BLOB_NAME, dt, (1, 3, INPUT_H, INPUT_W))

    stem_out = conv_bn_relu(
        network=network,
        weight_map=weight_map,
        lname="stem",
        input=data,
        out_ch=32,
        ksize=3,
        stride=2,
        padding=1,
    )

    # block 1
    x = add_resbottleneck_block(
        network,
        weight_map,
        lname="trunk_output.block1.block1-0",
        input=stem_out,
        in_ch=32,
        out_ch=32,
        stride=2,
        groups=2
    )

    # block 2
    for i in range(2):
        stride = 2 if i == 0 else 1
        x = add_resbottleneck_block(
            network, weight_map,
            lname=f"trunk_output.block2.block2-{i}",
            input=x,
            in_ch=32 if i == 0 else 64,
            out_ch=64,
            stride=stride,
            groups=4
        )

    # block 3
    for i in range(7):
        stride = 2 if i == 0 else 1
        x = add_resbottleneck_block(
            network, weight_map,
            lname=f"trunk_output.block3.block3-{i}",
            input=x,
            in_ch=64 if i == 0 else 160,
            out_ch=160,
            stride=stride,
            groups=10
        )

    # block 4
    for i in range(12):
        stride = 2 if i == 0 else 1
        x = add_resbottleneck_block(
            network, weight_map,
            lname=f"trunk_output.block4.block4-{i}",
            input=x,
            in_ch=160 if i == 0 else 400,
            out_ch=400,
            stride=stride,
            groups=25
        )

    pool = network.add_pooling_nd(
        input=x.get_output(0),
        window_size=trt.DimsHW(7, 7),
        type=trt.PoolingType.AVERAGE
    )
    pool.stride_nd = trt.DimsHW(7, 7)

    # fully connected.
    # ===== Flatten =====
    flatten = network.add_shuffle(pool.get_output(0))
    flatten.reshape_dims = (1, 400)        # -> (1, 400)

    # ===== FC weights =====
    weight_map["fc.weight"] = weight_map["fc.weight"].reshape(1000, 400)
    fc_weight_tensor = network.add_constant(
        weight_map["fc.weight"].shape,
        weight_map["fc.weight"].astype(np.float32)
    )

    # ===== MatrixMultiply (FC) =====
    fc_mm = network.add_matrix_multiply(
        input0=flatten.get_output(0),           # (1, 400)
        op0=trt.MatrixOperation.NONE,
        input1=fc_weight_tensor.get_output(0),  # (1000, 400)
        op1=trt.MatrixOperation.TRANSPOSE       # → (400, 1000)
    )

    # ===== Bias =====
    weight_map["fc.bias"] = weight_map["fc.bias"].reshape(1, 1000)
    fc_bias_tensor = network.add_constant(
        weight_map["fc.bias"].shape,
        weight_map["fc.bias"].astype(np.float32)
    )

    fc = network.add_elementwise(
        fc_mm.get_output(0),
        fc_bias_tensor.get_output(0),
        trt.ElementWiseOperation.SUM
    )

    fc.get_output(0).name = OUTPUT_BLOB_NAME
    network.mark_output(fc.get_output(0))

    # Build engine
    engine = builder.build_serialized_network(network, config)

    return engine


def API_to_model():
    builder = trt.Builder(TRT_LOGGER)
    config = builder.create_builder_config()
    engine = create_engine(builder, config, trt.float32)
    assert engine
    with open(ENGINE_PATH, "wb") as f:
        f.write(engine)


def do_inference(context, host_in, host_out, batchSize):
    engine = context.engine

    input_name = engine.get_tensor_name(0)
    output_name = engine.get_tensor_name(1)

    print("Input tensor name is :", input_name)
    print("output tensor name is :", output_name)

    device_in = cuda.mem_alloc(host_in.nbytes)
    device_out = cuda.mem_alloc(host_out.nbytes)

    stream = cuda.Stream()

    cuda.memcpy_htod_async(device_in, host_in, stream)

    context.set_tensor_address(input_name, int(device_in))
    context.set_tensor_address(output_name, int(device_out))

    context.execute_async_v3(stream_handle=stream.handle)

    cuda.memcpy_dtoh_async(host_out, device_out, stream)

    stream.synchronize()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-s", action='store_true')
    parser.add_argument("-d", action='store_true')
    args = parser.parse_args()

    if not (args.s ^ args.d):
        print(
            "arguments not right!\n"
            "python regnet.py -s   # serialize model to plan file\n"
            "python regnet.py -d   # deserialize plan file and run inference"
        )
        sys.exit()

    if args.s:
        API_to_model()
    else:
        runtime = trt.Runtime(TRT_LOGGER)
        assert runtime

        with open(ENGINE_PATH, "rb") as f:
            engine = runtime.deserialize_cuda_engine(f.read())
        assert engine

        context = engine.create_execution_context()
        assert context

        data = np.ones((BATCH_SIZE, 3, INPUT_H, INPUT_W), dtype=np.float32)
        host_in = cuda.pagelocked_empty(data.size, dtype=np.float32)
        np.copyto(host_in, data.ravel())
        host_out = cuda.pagelocked_empty(OUTPUT_SIZE, dtype=np.float32)

        do_inference(context, host_in, host_out, 1)

        print(f'Output: {host_out}')

File: regnet/regnet.cpp
#include <chrono>
#include <cmath>
#include <fstream>
#include <map>
#include "NvInfer.h"
#include "cuda_runtime_api.h"
#include "logging.h"

#define CHECK(status)                                          \
    do {                                                       \
        auto ret = (status);                                   \
        if (ret != 0) {                                        \
            std::cerr << "Cuda failure: " << ret << std::endl; \
            abort();                                           \
        }                                                      \
    } while (0)

// stuff we know about the network and the input/output blobs
static const int INPUT_H = 224;
static const int INPUT_W = 224;
static const int OUTPUT_SIZE = 1000;

const char* INPUT_BLOB_NAME = "data";
const char* OUTPUT_BLOB_NAME = "prob";

using namespace nvinfer1;

static Logger gLogger;

// Load weights from files shared with TensorRT samples.
// TensorRT weight files have a simple space delimited format:
// [type] [size] <data x size in hex>
std::map<std::string, Weights> loadWeights(const std::string file) {
    std::cout << "Loading weights: " << file << std::endl;
    std::map<std::string, Weights> weightMap;

    // Open weights file
    std::ifstream input(file);
    assert(input.is_open() && "Unable to load weight file.");

    // Read number of weight blobs
    int32_t count;
    input >> count;
    assert(count > 0 && "Invalid weight map file.");

    while (count--) {
        Weights wt{DataType::kFLOAT, nullptr, 0};
        uint32_t size;

        // Read name and type of blob
        std::string name;
        input >> name >> std::dec >> size;
        wt.type = DataType::kFLOAT;

        // Load blob
        uint32_t* val = reinterpret_cast<uint32_t*>(malloc(sizeof(val) * size));
        for (uint32_t x = 0, y = size; x < y; ++x) {
            input >> std::hex >> val[x];
        }
        wt.values = val;

        wt.count = size;
        weightMap[name] = wt;
    }

    return weightMap;
}

IScaleLayer* addBatchNorm2d(INetworkDefinition* network, std::map<std::string, Weights>& weightMap,
                            const std::string& lname, ITensor& input, float eps = 1e-5) {
    float* gamma = (float*)weightMap[lname + ".weight"].values;
    float* beta = (float*)weightMap[lname + ".bias"].values;
    float* mean = (float*)weightMap[lname + ".running_mean"].values;
    float* var = (float*)weightMap[lname + ".running_var"].values;

    int len = weightMap[lname + ".running_var"].count;
    std::cout << "[BatchNorm] " << lname << " channels = " << len << std::endl;

    // scale = gamma / sqrt(var + eps)
    float* scaleVal = reinterpret_cast<float*>(malloc(sizeof(float) * len));
    for (int i = 0; i < len; ++i) {
        scaleVal[i] = gamma[i] / sqrt(var[i] + eps);
    }
    Weights scale{DataType::kFLOAT, scaleVal, len};

    // shift = beta - mean * scale
    float* shiftVal = reinterpret_cast<float*>(malloc(sizeof(float) * len));
    for (int i = 0; i < len; ++i) {
        shiftVal[i] = beta[i] - mean[i] * scaleVal[i];
    }
    Weights shift{DataType::kFLOAT, shiftVal, len};

    // power = 1.0
    float* powerVal = reinterpret_cast<float*>(malloc(sizeof(float) * len));
    for (int i = 0; i < len; ++i) {
        powerVal[i] = 1.0f;
    }
    Weights power{DataType::kFLOAT, powerVal, len};

    // save in weightMap in case of memory leak.
    weightMap[lname + ".scale"] = scale;
    weightMap[lname + ".shift"] = shift;
    weightMap[lname + ".power"] = power;

    IScaleLayer* bn = network->addScale(input, ScaleMode::kCHANNEL, shift, scale, power);
    assert(bn);

    return bn;
}

IActivationLayer* convBnRelu(INetworkDefinition* network, std::map<std::string, Weights>& weightMap,
                             const std::string& lname, ITensor& input, int out_ch, int ksize, int stride, int padding,
                             int groups = 1) {
    // 1. conv
    std::string convWeightName = lname + ".0.weight";
    IConvolutionLayer* conv =
            network->addConvolutionNd(input, out_ch, DimsHW{ksize, ksize}, weightMap[convWeightName], Weights{});
    assert(conv);

    conv->setStrideNd(DimsHW{stride, stride});
    conv->setPaddingNd(DimsHW{padding, padding});
    conv->setNbGroups(groups);

    // 2. BatchNorm
    std::string bnName = lname + ".1";
    IScaleLayer* bn = addBatchNorm2d(network, weightMap, bnName, *conv->getOutput(0), 1e-5f);
    assert(bn);

    // 3. relu
    IActivationLayer* relu = network->addActivation(*bn->getOutput(0), ActivationType::kRELU);
    assert(relu);

    return relu;
}

IActivationLayer* addResBottleneckBlock(INetworkDefinition* network, std::map<std::string, Weights>& weightMap,
                                        const std::string& lname, ITensor& input, int in_ch, int out_ch, int stride,
                                        int groups) {
    ITensor* in_tensor = &input;

    // 1. calculate bottleneck channel.
    // f.a conv weight shape = [bottleneck, in_ch, 1, 1]
    float* wa = (float*)weightMap[lname + ".f.a.0.weight"].values;
    int wa_count = weightMap[lname + ".f.a.0.weight"].count;

    int bottleneck = wa_count / in_ch;  // 1x1 conv
    std::cout << "[Block] " << lname << " bottleneck = " << bottleneck << std::endl;

    // f.c out_ch
    float* wc = (float*)weightMap[lname + ".f.c.0.weight"].values;
    int wc_count = weightMap[lname + ".f.c.0.weight"].count;

    out_ch = wc_count / bottleneck;
    std::cout << "[Block] " << lname << " out_ch = " << out_ch << std::endl;

    // 2. projection branch
    ITensor* proj_out = nullptr;

    if (stride != 1 || in_ch != out_ch) {
        // 1×1 conv
        IConvolutionLayer* proj_conv = network->addConvolutionNd(*in_tensor, out_ch, DimsHW{1, 1},
                                                                 weightMap[lname + ".proj.0.weight"], Weights{});
        assert(proj_conv);
        proj_conv->setStrideNd(DimsHW{stride, stride});

        // BN
        IScaleLayer* proj_bn = addBatchNorm2d(network, weightMap, lname + ".proj.1", *proj_conv->getOutput(0));
        proj_out = proj_bn->getOutput(0);
    } else {
        // identity
        proj_out = in_tensor;
    }

    // 3. f.a 1×1 conv + BN + ReLU
    IActivationLayer* a = convBnRelu(network, weightMap, lname + ".f.a", *in_tensor, bottleneck,
                                     1,  // ksize
                                     1,  // stride
                                     0   // padding
    );

    // 4. f.b 3×3 conv (groups) + BN + ReLU
    IActivationLayer* b = convBnRelu(network, weightMap, lname + ".f.b", *a->getOutput(0), bottleneck,
                                     3,       // ksize
                                     stride,  // stride (block stride)
                                     1,       // padding
                                     groups);

    // 5. f.c 1×1 conv + BN
    IConvolutionLayer* c_conv = network->addConvolutionNd(*b->getOutput(0), out_ch, DimsHW{1, 1},
                                                          weightMap[lname + ".f.c.0.weight"], Weights{});
    assert(c_conv);

    IScaleLayer* c_bn = addBatchNorm2d(network, weightMap, lname + ".f.c.1", *c_conv->getOutput(0));

    // 6. elementwise sum
    IElementWiseLayer* ew = network->addElementWise(*c_bn->getOutput(0), *proj_out, ElementWiseOperation::kSUM);
    assert(ew);

    // 7. ReLU
    IActivationLayer* out = network->addActivation(*ew->getOutput(0), ActivationType::kRELU);
    assert(out);

    return out;
}

ICudaEngine* createEngine(unsigned int maxBatchSize, IBuilder* builder, IBuilderConfig* config, DataType dt) {
    // 1. Create Network
    INetworkDefinition* network =
            builder->createNetworkV2(1U << (unsigned int)NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
    assert(network);

    // 2. Input
    ITensor* data = network->addInput(INPUT_BLOB_NAME, dt, Dims4{1, 3, INPUT_H, INPUT_W});
    assert(data);
    std::cout << "Input shape = (1,3," << INPUT_H << "," << INPUT_W << ")" << std::endl;

    // 3. Load weights
    std::map<std::string, Weights> weightMap = loadWeights("../models/regnet_x_400mf.wts");

    std::cout << "Loaded " << weightMap.size() << " weights." << std::endl;

    // 4. Stem  (Conv-BN-ReLU)
    // stem: 3×3, stride=2, out=32
    IActivationLayer* stem = convBnRelu(network, weightMap, "stem", *data,
                                        32,  // out_ch
                                        3,   // ksize
                                        2,   // stride
                                        1    // padding
    );

    // 5. Block1 (1 block)
    // channels: 32 → 32, stride=2, groups=2
    std::cout << "----- Block1 -----" << std::endl;
    IActivationLayer* x = addResBottleneckBlock(network, weightMap, "trunk_output.block1.block1-0", *stem->getOutput(0),
                                                32,  // in_ch
                                                32,  // out_ch
                                                2,   // stride
                                                2    // groups
    );

    // 6. Block2 (2 blocks)
    // channels: 32 → 64, stride=2, groups=4
    std::cout << "----- Block2 -----" << std::endl;
    for (int i = 0; i < 2; i++) {
        int stride = (i == 0 ? 2 : 1);
        int in_ch = (i == 0 ? 32 : 64);

        x = addResBottleneckBlock(network, weightMap, "trunk_output.block2.block2-" + std::to_string(i),
                                  *x->getOutput(0), in_ch, 64, stride,
                                  4  // groups
        );
    }

    // 7. Block3 (7 blocks)
    // channels: 64 → 160, stride=2, groups=10
    std::cout << "----- Block3 -----" << std::endl;
    for (int i = 0; i < 7; i++) {
        int stride = (i == 0 ? 2 : 1);
        int in_ch = (i == 0 ? 64 : 160);

        x = addResBottleneckBlock(network, weightMap, "trunk_output.block3.block3-" + std::to_string(i),
                                  *x->getOutput(0), in_ch, 160, stride,
                                  10  // groups
        );
    }

    // 8. Block4 (12 blocks)
    // channels: 160 → 400, stride=2, groups=25
    std::cout << "----- Block4 -----" << std::endl;
    for (int i = 0; i < 12; i++) {
        int stride = (i == 0 ? 2 : 1);
        int in_ch = (i == 0 ? 160 : 400);

        x = addResBottleneckBlock(network, weightMap, "trunk_output.block4.block4-" + std::to_string(i),
                                  *x->getOutput(0), in_ch, 400, stride,
                                  25  // groups
        );
    }

    // 9. AvgPool 7×7 (stride=7)
    IPoolingLayer* pool = network->addPoolingNd(*x->getOutput(0), PoolingType::kAVERAGE, DimsHW{7, 7});
    assert(pool);
    pool->setStrideNd(DimsHW{7, 7});

    std::cout << "After avg pool: " << pool->getOutput(0)->getDimensions().d[0] << ","
              << pool->getOutput(0)->getDimensions().d[1] << "," << pool->getOutput(0)->getDimensions().d[2]
              << std::endl;

    // 10. Flatten
    IShuffleLayer* flatten = network->addShuffle(*pool->getOutput(0));
    flatten->setReshapeDimensions(Dims2{1, 400});  // shape -> (1, 400)
    assert(flatten);

    // 11. FC Layer (MatrixMultiply + Bias)
    // FC weight: [1000, 400]
    // reshape fc weight
    Weights& fcw = weightMap["fc.weight"];
    std::cout << "fc weight count = " << fcw.count << std::endl;

    // constant for FC weight
    IConstantLayer* fc_weight_tensor = network->addConstant(Dims2{1000, 400}, fcw);
    assert(fc_weight_tensor);

    // MatMul: (1×400) × (400×1000) → (1×1000)
    IMatrixMultiplyLayer* fc_mm =
            network->addMatrixMultiply(*flatten->getOutput(0), MatrixOperation::kNONE, *fc_weight_tensor->getOutput(0),
                                       MatrixOperation::kTRANSPOSE);
    assert(fc_mm);

    // Bias (1×1000)
    IConstantLayer* fc_bias_tensor = network->addConstant(Dims2{1, 1000}, weightMap["fc.bias"]);
    assert(fc_bias_tensor);

    IElementWiseLayer* fc =
            network->addElementWise(*fc_mm->getOutput(0), *fc_bias_tensor->getOutput(0), ElementWiseOperation::kSUM);
    assert(fc);

    // 12. Output
    fc->getOutput(0)->setName(OUTPUT_BLOB_NAME);
    network->markOutput(*fc->getOutput(0));

    // 13. Build engine
    ICudaEngine* engine = builder->buildEngineWithConfig(*network, *config);
    assert(engine);

    std::cout << "Engine build completed!" << std::endl;

    return engine;
}

void APIToModel(unsigned int maxBatchSize, IHostMemory** modelStream) {
    // Create builder
    IBuilder* builder = createInferBuilder(gLogger);
    IBuilderConfig* config = builder->createBuilderConfig();

    // Create model to populate the network, then set the outputs and create an engine
    ICudaEngine* engine = createEngine(maxBatchSize, builder, config, DataType::kFLOAT);
    assert(engine != nullptr);

    // Serialize the engine
    (*modelStream) = engine->serialize();
}

void doInference(IExecutionContext& context, float* input, float* output, float* pool3FlattenLayerOutput,
                 int batchSize) {
    const ICudaEngine& engine = context.getEngine();

    const char* inputName = INPUT_BLOB_NAME;
    const char* outputName = OUTPUT_BLOB_NAME;

    void* deviceInput{nullptr};
    void* deviceOutput{nullptr};

    // Create GPU buffers on device
    CHECK(cudaMalloc(&deviceInput, batchSize * 3 * INPUT_H * INPUT_W * sizeof(float)));
    CHECK(cudaMalloc(&deviceOutput, batchSize * OUTPUT_SIZE * sizeof(float)));

    // Create stream
    cudaStream_t stream;
    CHECK(cudaStreamCreate(&stream));

    // DMA input batch data to device, infer on the batch asynchronously, and DMA output back to host
    CHECK(cudaMemcpyAsync(deviceInput, input, batchSize * 3 * INPUT_H * INPUT_W * sizeof(float), cudaMemcpyHostToDevice,
                          stream));

    context.setTensorAddress(inputName, deviceInput);
    context.setTensorAddress(outputName, deviceOutput);

    context.enqueueV3(stream);

    CHECK(cudaMemcpyAsync(output, deviceOutput, batchSize * OUTPUT_SIZE * sizeof(float), cudaMemcpyDeviceToHost,
                          stream));

    cudaStreamSynchronize(stream);

    // Release stream and buffers
    cudaStreamDestroy(stream);
    CHECK(cudaFree(deviceInput));
    CHECK(cudaFree(deviceOutput));
}

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cerr << "arguments not right!" << std::endl;
        std::cerr << "./regnet -s   // serialize model to plan file" << std::endl;
        std::cerr << "./regnet -d   // deserialize plan file and run inference" << std::endl;
        return -1;
    }

    // create a model using the API directly and serialize it to a stream
    char* trtModelStream{nullptr};
    size_t size{0};

    if (std::string(argv[1]) == "-s") {
        IHostMemory* modelStream{nullptr};
        APIToModel(1, &modelStream);
        assert(modelStream != nullptr);

        std::ofstream p("../models/regnet_x_400mf.engine", std::ios::binary);
        if (!p) {
            std::cerr << "could not open plan output file" << std::endl;
            return -1;
        }
        p.write(reinterpret_cast<const char*>(modelStream->data()), modelStream->size());
        return 1;
    } else if (std::string(argv[1]) == "-d") {
        std::ifstream file("../models/regnet_x_400mf.engine", std::ios::binary);
        if (file.good()) {
            file.seekg(0, file.end);
            size = file.tellg();
            file.seekg(0, file.beg);
            trtModelStream = new char[size];
            assert(trtModelStream);
            file.read(trtModelStream, size);
            file.close();
        }
    } else {
        return -1;
    }

    // Subtract mean from image
    float data[3 * INPUT_H * INPUT_W];
    for (int i = 0; i < 3 * INPUT_H * INPUT_W; i++)
        data[i] = 1;

    IRuntime* runtime = createInferRuntime(gLogger);
    assert(runtime != nullptr);
    ICudaEngine* engine = runtime->deserializeCudaEngine(trtModelStream, size);
    assert(engine != nullptr);
    IExecutionContext* context = engine->createExecutionContext();
    assert(context != nullptr);

    float prob[OUTPUT_SIZE];
    float pool3FlattenLayerOutput[9216];
    doInference(*context, data, prob, pool3FlattenLayerOutput, 1);

    // Print histogram of the output distribution
    std::cout << "\nOutput:\n\n";
    for (unsigned int i = 0; i < OUTPUT_SIZE; i++) {
        std::cout << prob[i] << ", ";
        if (i % 10 == 0)
            std::cout << i / 10 << std::endl;
    }
    std::cout << std::endl;

    return 0;
}
